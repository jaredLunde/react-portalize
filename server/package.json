{
  "name": "react-portalize-server",
  "version": "1.0.0",
  "homepage": "https://github.com/jaredLunde/react-portalize#readme",
  "repository": "github:jaredLunde/react-portalize",
  "bugs": "https://github.com/jaredLunde/react-portalize/issues",
  "license": "MIT",
  "main": "dist/main/index.js",
  "module": "dist/module/index.js",
  "source": "src/index.ts",
  "types": "src/index.d.ts",
  "files": [
    "/src",
    "/dist"
  ],
  "sideEffects": false
}
