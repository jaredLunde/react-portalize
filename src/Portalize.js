import React from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'


/**
<Portalize container='#portals'>
  <div>Inject me</div>
</Portalize>
*/
export const PORTALS = {}

export default class Portalize extends React.Component {
  static propTypes = {
    container: PropTypes.string.isRequired
  }

  static defaultProps = {
    container: '#portals'
  }

  nodes = null

  render () {
    this.nodes =
      typeof document !== 'undefined' && document.querySelectorAll(this.props.container)

    if (this.nodes === false) {
      PORTALS[this.props.container] = this.props.children
      return null
    }
    else if (this.nodes.length === 0) {
      return null
    }
    else {
      const portals = []

      for (let i = 0; i < this.nodes.length; i++) {
        portals.push(ReactDOM.createPortal(this.props.children, this.nodes[i]))
      }

      return portals
    }
  }
}
