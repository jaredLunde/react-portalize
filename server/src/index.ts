import {renderToStaticMarkup} from 'react-dom/server'
import {load} from 'cheerio'
import {PORTALS} from 'react-portalize'

export function renderPortalsToString(html: string): string {
  const selectors = Object.keys(PORTALS)

  if (selectors.length === 0) return html

  const doc = load(html)

  for (let i = 0; i < selectors.length; i++) {
    const selector = selectors[i]
    doc(selector).html(renderToStaticMarkup(PORTALS[selector]))
  }

  return String(
    html.match(/^\s*(<!|<html|<body|<head)/)
      ? doc.html()
      : doc('body').contents()
  )
}
