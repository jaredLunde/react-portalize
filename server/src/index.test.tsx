/**
 * @jest-environment node
 */
import * as React from 'react'
import {renderToStaticMarkup, renderToString} from 'react-dom/server'
import Portalize, {PORTALS} from '../../src/index'
import {renderPortalsToString} from './index'

const Context = React.createContext({foo: null})

const Component = () => {
  const context = React.useContext(Context)
  return context.foo
}

describe('<Portalize>', () => {
  it('should store portals in the PORTALS constant according to their container', () => {
    const result = renderToString(
      <body>
        <div id="portals" />
        <Portalize container=".portals">
          <div>Hello world</div>
        </Portalize>
      </body>
    )

    expect(Object.values(PORTALS).length).toBe(1)
    expect(PORTALS['.portals'].props.children).toBe('Hello world')
    expect(result).toMatchSnapshot()
    delete PORTALS['.portals']
  })

  it('should work with context', () => {
    const result = renderToString(
      <body>
        <div id="portals" />
        <Portalize
          container=".portals"
          providers={[{provider: Context.Provider, value: {foo: 'bar'}}]}
        >
          <Component />
        </Portalize>
      </body>
    )

    expect(Object.values(PORTALS).length).toBe(1)
    expect(renderToStaticMarkup(PORTALS['.portals'])).toMatchSnapshot()
    expect(result).toMatchSnapshot()
    delete PORTALS['.portals']
  })

  it('should not render on server if server={false}', () => {
    const result = renderToString(
      <body>
        <div id="portals" />
        <Portalize container=".portals" server={false}>
          <div>Hello world</div>
        </Portalize>
      </body>
    )

    expect(Object.values(PORTALS).length).toBe(0)
    expect(result).toMatchSnapshot()
  })

  it('should not render portals to a string', () => {
    const result = renderPortalsToString(
      renderToString(
        <body>
          <div id="portals" />
          <Portalize container=".portals">
            <div>Hello world</div>
          </Portalize>
        </body>
      )
    )

    expect(result).toMatchSnapshot()
  })
})
