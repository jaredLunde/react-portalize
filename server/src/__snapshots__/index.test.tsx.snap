// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portalize> should not render on server if server={false} 1`] = `"<body data-reactroot=\\"\\"><div id=\\"portals\\"></div></body>"`;

exports[`<Portalize> should not render portals to a string 1`] = `"<body data-reactroot=\\"\\"><div id=\\"portals\\"></div></body>"`;

exports[`<Portalize> should store portals in the PORTALS constant according to their container 1`] = `"<body data-reactroot=\\"\\"><div id=\\"portals\\"></div></body>"`;

exports[`<Portalize> should work with context 1`] = `"bar"`;

exports[`<Portalize> should work with context 2`] = `"<body data-reactroot=\\"\\"><div id=\\"portals\\"></div></body>"`;
